src/Type.h

```diff
@@ -20,81 +20,92 @@ struct Expr;
  * types. Instead vectorize a function. */
 struct Type {
-    /** The basic type code: signed integer, unsigned integer, or floating point.
-     *
-     * Note that TypeCode is guaranteed to have values identical to those of
-     * halide_type_code_t (HalideRuntime.h), but exists as a separate typedef
-     * to preserve source code compatibility.
-     */
-    enum TypeCode {
-        Int = halide_type_int,
-        UInt = halide_type_uint,
-        Float = halide_type_float,
-        Handle = halide_type_handle
-    } code;
-
-    /** The number of bits of precision of a single scalar value of this type. */
-    int bits;
+  private:
+    halide_type_t type;
 
-    /** The number of bytes required to store a single scalar value of this type. Ignores vector width. */
-    int bytes() const {return (bits + 7) / 8;}
+  public:
+    /** Aliases for halide_type_code_t values for legacy compatibility
+     * and to match the Halide internal C++ style. */
+    // @{
+    static constexpr halide_type_code_t Int = halide_type_int;
+    static constexpr halide_type_code_t UInt = halide_type_uint;
+    static constexpr halide_type_code_t Float = halide_type_float;
+    static constexpr halide_type_code_t Handle = halide_type_handle;
+    // @}
 
-    /** How many elements (if a vector type). Should be 1 for scalar types. */
-    int width;
+    /** The number of bytes required to store a single scalar value of this type. Ignores vector width. */
+    int bytes() const {return (bits() + 7) / 8;}
 
     // Default ctor initializes everything to predictable-but-unlikely values
-    Type() : code(Handle), bits(0), width(0) {}
+    Type() : type(Handle, 0, 0) {}
+
+    Type(halide_type_code_t code, uint8_t bits, int width) : type(code, (uint8_t)bits, (uint16_t)width) {}
 
-    Type(TypeCode _code, int _bits, int _width) : code(_code), bits(_bits), width(_width) {}
+    Type(const Type &that) : type(that) {}
+
+    Type(const halide_type_t &that) : type(that) {}
+
+    operator halide_type_t() const { return type; }
+
+    halide_type_code_t code() const { return type.code; }
+    int bits() const { return type.bits; }
+    int width() const { return type.width; }
+
+    /** Return Type with same number of bits and width, but new_codefor a type code. */
+    Type with_code(halide_type_code_t new_code) const {
+        return Type(new_code, bits(), width());
+    }
 
-    Type(const Type &that) : code(that.code), bits(that.bits), width(that.width) {}
+    /** Return Type with same type code and width, but new_bits for the number of bits. */
+    Type with_bits(uint8_t new_bits) const {
+        return Type(code(), new_bits, width());
+    }
 
-    Type& operator=(const Type& that) {
-        this->code = that.code;
-        this->bits = that.bits;
-        this->width = that.width;
-        return *this;
+    /** Return Type with same type code and number of bits, but new_width for the vector width. */
+    // TODO(zalman): Should this be called broadcast?
```

Comment:
I think vector_of is more descriptive, but it's also not great because the name suggests you'll get a vector of vectors if called on a type that is already a vector. If we kept that name, we should add an assert that the current vector width is 1.

While we're at this change, I've been briefly confused in the past when I attempted to use "width" to mean "integer width" and not "vector width". Maybe width should be vector_width? i.e. with_vector_width.

That was a bit of a ramble. To clarify, I think we should either keep vector_of, but add an assert that the type is scalar, or we should keep with_width, but consider renaming it to with_vector_width (and same for width()).